libkbfs/backpressure_disk_limiter.go

```diff
@@ -69,71 +83,98 @@ func defaultDoDelay(ctx context.Context, delay time.Duration) error {
-	return newBackpressureDiskLimiterWithDelayFunction(
+	backpressureMinThreshold, backpressureMaxThreshold float64,
+	byteLimit int64, maxDelay time.Duration,
+	journalPath string) *backpressureDiskLimiter {
+	return newBackpressureDiskLimiterWithFunctions(
 		backpressureMinThreshold, backpressureMaxThreshold,
-		byteLimit, maxDelay, defaultDoDelay)
+		byteLimit, maxDelay, defaultDoDelay, func() (int64, error) {
+			availBytes, err := getDiskLimits(journalPath)
+			if err != nil {
+				return 0, err
+			}
+
+			if availBytes > uint64(math.MaxInt64) {
+				return math.MaxInt64, nil
+			}
+			return int64(availBytes), nil
+		})
 }
 
-func (s backpressureDiskLimiter) onJournalEnable(journalBytes int64) int64 {
-	if journalBytes == 0 {
-		return s.s.Count()
-	}
-	return s.s.ForceAcquire(journalBytes)
+func (s *backpressureDiskLimiter) onJournalEnable(journalBytes int64) int64 {
+	s.journalBytesLock.Lock()
+	defer s.journalBytesLock.Unlock()
+	s.journalBytes += journalBytes
+	return 0
 }
 
-func (s backpressureDiskLimiter) onJournalDisable(journalBytes int64) {
-	if journalBytes > 0 {
-		s.s.Release(journalBytes)
-	}
+func (s *backpressureDiskLimiter) onJournalDisable(journalBytes int64) {
+	s.journalBytesLock.Lock()
+	defer s.journalBytesLock.Unlock()
+	s.journalBytes -= journalBytes
 }
 
-func (s backpressureDiskLimiter) getDelay() time.Duration {
-	availBytes := s.s.Count()
-	usedBytes := s.byteLimit - availBytes
-	if usedBytes <= s.backpressureMinThreshold {
-		return 0
-	}
-
-	if usedBytes >= s.backpressureMaxThreshold {
-		return s.maxDelay
-	}
+func boundedLerp(x, min, max float64) float64 {
+	return math.Min(1.0, math.Max(0.0, (x-min)/(max-min)))
+}
 
-	scale := float64(usedBytes-s.backpressureMinThreshold) /
-		float64(s.backpressureMaxThreshold-s.backpressureMinThreshold)
-	delayNs := int64(float64(s.maxDelay.Nanoseconds()) * scale)
-	return time.Duration(delayNs) * time.Nanosecond
+func (s *backpressureDiskLimiter) getDelay(availBytes int64) time.Duration {
+	journalBytes := func() int64 {
+		s.journalBytesLock.RLock()
+		defer s.journalBytesLock.RUnlock()
+		return s.journalBytes
+	}()
+
+	r := math.Max(
+		float64(journalBytes)/float64(s.maxJournalBytes),
+		float64(journalBytes)/(float64(journalBytes)+float64(availBytes)))
+	m := s.backpressureMinThreshold
+	M := s.backpressureMaxThreshold
+	scale := math.Min(1.0, math.Max(0.0, (r-m)/(M-m)))
+	return time.Duration(scale * float64(s.maxDelay))
 }
 
-func (s backpressureDiskLimiter) beforeBlockPut(
+func (s *backpressureDiskLimiter) beforeBlockPut(
 	ctx context.Context, blockBytes int64,
 	log logger.Logger) (int64, error) {
 	if blockBytes == 0 {
 		// Better to return an error than to panic in Acquire.
-		return s.s.Count(), errors.New(
+		return 0, errors.New(
 			"beforeBlockPut called with 0 blockBytes")
 	}
 
-	delay := s.getDelay()
+	// We don't actually look at blockBytes -- we're assuming that
+	// it's small compared to the other numbers.
+
+	availBytes, err := s.availBytesFn()
```

Comment:
Do you think this is too expensive to call on every block put? If so, I can do something like only update this every n seconds, and in between approximate it by subtracting put blocks in afterBlockPut.